### DISCUSSION TITLE
Sending single key to the shell when with a modified

### DISCUSSION THREAD

**Discussion by liaowang11** [ghostty-org/ghostty#3502] [Feature Requests, Ideas] [answered]
This bug is quite hard to describe and I'm not sure whether it's ghostty or fish shell, but at least it's not happening in kitty.
I'm currently using fish shell with hybrid keybindings, which means that `left_bracket` in normal mode maps to `history-token-search-backward`. And in insert mode, I like to press `Ctrl+left_brakcet` to exit insert mode.

When I press `Ctrl+left_brakcet`, the previous history item flashes and get cleared immediately. Ghostty seems to be sending `left_bracket ` alone to fish shell.

The video uploaded may do better than my description.

https://github.com/user-attachments/assets/aefb9d77-337e-42e6-95a0-de31180e2e6b

**Comment by mitchellh** [comment]
Convert to discussion for issue triage.

**Comment by liaowang11** [comment]
Using `fish_key_reader` and input `Ctrl-left_bracket` in kitty gets:
<img width="184" alt="image" src="https://github.com/user-attachments/assets/41de7251-8f26-4752-8b2d-21af729b0df4" />
But in ghostty:
<img width="291" alt="image" src="https://github.com/user-attachments/assets/5a029683-9984-4560-b78e-e8e6460cf173" />

**Reply by 00-kat** [reply]
Likely the fault of [fixterms], see https://github.com/ghostty-org/ghostty/discussions/3151#discussioncomment-11678064.  TL;DR: Ghostty implements [fixterms], which many other terminal emulators don't; you can either bind `\e\[91\;5u` to do what you want in Fish, or rebind `ctrl+left_bracket` to send the same sequence as Kitty.

Note: Fish since version 4 (which is still in beta right now) supports Fixterms, so it'll just work™ there.

[fixterms]: https://www.leonerd.org.uk/hacks/fixterms

**Reply by liaowang11** [reply]
Using fish beta, the problem is still there, while `fish_key_reader` returns the same.
```
❯ fish --version
fish, version 4.0b1
```
Ghostty:
<img width="212" alt="image" src="https://github.com/user-attachments/assets/003afb59-d518-43ab-8968-0ce89ea46408" />
Iterm2:
<img width="223" alt="image" src="https://github.com/user-attachments/assets/4ecf755e-be8f-4688-b0d1-48cfc020ee1a" />

**Reply by 00-kat** [reply, marked as answer]
Ohh apologies I misinterpreted.

So for some context: `ctrl+[` and escape encode to the same thing, and Fish binds escape to exit insert mode but doesn't bind `ctrl+[` to escape insert mode.  But since they encode to the same thing, Fish reads `ctrl+[` as escape.  Ghostty implements fixterms, which changes the two encodings to something different.  Fish since version 4 supports fixterms so it interprets that weird `\e\[91\;5u` thing (A.K.A. a CSI-u encoded key sequence) as `ctrl+[`, which is *correct*, but you actually want it to be interpreted as escape, and hence escape insert mode.

So the following solutions are possible:
  - If you want to use Fish before version 4, add the following to [`fish_user_key_bindings`] or your `config.fish`[^1]:
    ```sh
    bind -M insert \e\[91\;5u 'set fish_bind_mode default; commandline -f backward-char repaint-mode'
    ```
  - If you want to use Fish version 4 or later, add the following to [`fish_user_key_bindings`] or your `config.fish`[^1]:
    ```sh
    bind -M insert 'ctrl-[' 'set fish_bind_mode default; commandline -f backward-char repaint-mode'
    ```
  - If you want `ctrl+[` to work as escape in all applications (e.g. Fish, Vim, etc.), add the following to your Ghostty configuration:
    ```ini
    keybind = ctrl+left_bracket=text:\x1b
    ```

[`fish_user_key_bindings`]: https://fishshell.com/docs/current/interactive.html#custom-bindings

[^1]: Note: I use `set fish_bind_mode default; commandline -f backward-char repaint-mode` here, while the default binding for escape is actually `if commandline -P; commandline -f cancel; else; set fish_bind_mode default; commandline -f backward-char repaint-mode; end'`; I'll let you decide which you want (the default binding for escape will escape the pager menu from completions and the like if it's open, and will otherwise switch to normal mode, while what I have will always switch to normal mode).

**Reply by liaowang11** [reply]
Sorry, but I still don't understand. On my machine *without* the tweaks you provided(maybe related to me using hybrid mode?), `Ctrl-[` works in insert mode to escape to normal mode, the issue is the **flash of content during the switch from insert mode back to normal mode**. 

Using the first two solutions, it's not solved, but the last one do. Maybe it's related to the difference between the actual `escape` commands and the commands in the tweak.

Anyway, thanks for your thorough explanation.

**Reply by liaowang11** [reply]
My mistake, I'm actually running fish 3.7 for system shell. This works like a charm. Thanks again.

**Reply by 00-kat** [reply]
Based on your second message, I'm guessing that it did work?

> **flash of content during the switch from insert mode back to normal mode**

Not sure why this is happening.  Fish does this when you try to complete something like `echo somethingthatdefinitelycantbecompleted`, and there are no matches.  There are probably a few other cases where it does it too, and I'm guessing that interpreting `\e[91;5u` as keypresses is causing it to trigger some action which makes it flash, before eventually switching to normal mode after a timeout or something (likely because of the `\e`, which is an escape).

**Reply by liaowang11** [reply]
Yes it works.
